Fix 2701 too may files found

sauliusg opened the pull request:
<!-- describe the changes you have made here: what, why, ... -->

I would like to offer my fix for the issue  2701: JabRef would find too many files when they are search with the [year]_[auth].* pattern. The reason seems to be a 'null' database passed to the BibtexKeyPatternUtil.makeLabel(BibEntry, String, Character, BibDatabase) method, and a null pointer exception is triggered. As a result, the '[auth]' part of the pattern was erased and ignored. With the suggested fix, JabRef uses the '[auth]' part of the pattern and finds just the files with the matching author, as expected. A relevant unit test was added; no other tests were broken.

- [ ] Change in CHANGELOG.md described -- no functionality changes, restored the previous state
- [x ] Tests created for changes -- 1 unit test added;
- [ ] Screenshots added (for bigger UI changes) -- no GUI changes;
- [ ] Manually tested changed features in running JabRef -- none
- [x ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?) -- the issue #2701 still open.
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`? -- no localisation changes.

Siedlerchr commented:
Thank you very much for your contribution 👍 . Codewise it looks very good to me. If the test passes and another dev gives his okay, we can merge it.

lynyus commented:
Thanks for the contribution @sauliusg . Great, when users find an issue and even fix it themselves (;

> The reason seems to be a 'null' database passed to the BibtexKeyPatternUtil.makeLabel(BibEntry, String, Character, BibDatabase) method, and a null pointer exception is triggered

I will investigate this tomorrow, because I'm not entirely sure if your fix solves the symptoms or we should dig a bit deeper and eliminate the main problem, which could be that bloody 'null' database.

sauliusg commented:
Dear  Siedlerchr, dear Lynyus,

thank you for the instant reply!

I have got feedback from Codacy and fixed: removed an unused variable from the new unit test. Also, I added another second unit test to cover the second 'if' statement added by me.

> I will investigate this tomorrow, because I'm not entirely sure if your fix solves the symptoms or we 
> should dig a bit deeper and eliminate the main problem, which could be that bloody 'null' database.

Please have a look. It seems to me that the 'null' can be traced to the explicit 'null' argument passed to it.

Regards,
Saulius

tobiasdiez commented:
@lynyus your suggestions sounds reasonable. We should try to not pass null as an argument. Moreover, the label generation code probably needs a refactoring (especially, get rid of the huge amount of static methods). But this will be probably a bigger project. While I would welcome PRs from you or @sauliusg in this direction, I think the current PR is an acceptable workaround within the boundaries of the current code infrastructure.

tobiasdiez commented on src/test/java/org/jabref/logic/bibtexkeypattern/MakeLabelWithoutDatabaseTest.java:
```diff
@@ -0,0 +1,46 @@
+    @Test
+    public void makeLabelForFileSearch() {
+        String label = BibtexKeyPatternUtil.makeLabel(entry,
+                /*value=*/ "auth",
```
I think we never used comments to indicate the name of the argument and thus would prefer that you remove them.

sauliusg replied:
Hmmm, indicating “Named Function Parameter” Comments is a recommended practice from the Dustin Boswell, Trevor Foucher "The Art of Readable Code", for languages that do not support named parameters, and I personally find such code easier to read, with not so much overhead. But if this is not used on JabRef and you do not find them useful, then, sure, please just remove them. Should I make the requested changes and recommit, or will you andle these issues?

lynyus replied:
Yes please, IntelliJ, for example automatically displays the parameter names. 
![parameter_names](https://cloud.githubusercontent.com/assets/1254003/25013947/0f762be8-2076-11e7-8699-e3e49089058e.png)

sauliusg replied:
Both issues fixed, please find a new push.

tobiasdiez commented on src/test/java/org/jabref/logic/bibtexkeypattern/MakeLabelWithoutDatabaseTest.java:
```diff
@@ -0,0 +1,46 @@
+
+public class MakeLabelWithoutDatabaseTest {
+
+    // private GlobalBibtexKeyPattern pattern;
```
Use remove the unused variable completely.

sauliusg replied:
Agreed, this var will probably never be needed.

Siedlerchr commented:
Okay, lgtm! I merge it in! Thank you very much for your contribution! 👍  Feel free to take a look at other issues

sauliusg commented:
Thanks!
